File: Algorithm-Templates/LeetCode-Daily/0003-LongestSubstring/LongestSubstring.c
#include <stdio.h>

int lengthOfLongestSubstring(char* s) {
    int last[256];
    //初始化数组
    for (int i = 0; i < 256; i++) {
        last[i] = -1;
    }
    int left = 0;
    int ans = 0;//符合条件的子串的长度
    for (int right = 0; s[right] != '\0'; right++) {
        unsigned char c = (unsigned char)s[right];//按无符号数解释，防止出现负数
        int prev = last[c];//prev表示当前字符上一次出现的位置
        if (prev >= left) {//如果上一次出现在窗口外
            left = prev + 1;//则更新窗口左边界
        }
        last[c] = right;//更新当前字符最新出现的位置
        int len = right - left + 1;//计算长度
        if (len > ans) {//ans表示最大长度
            ans = len;
        }
    }
    return ans;//返回的是长度
}

int main(void) {
    const char* cases[] = {
        "abcabcbb",
        "bbbbb",
        "pwwkew",
        "",
    };
    int expected[] = {3, 1, 3, 0};//预期的正确值
    int n = (int)(sizeof(cases) / sizeof(cases[0]));//计算出测试用例中共有多少字符串
    for (int i = 0; i < n; i++) {
        int out = lengthOfLongestSubstring((char*)cases[i]);//这里是强制类型转换，把只读的指针改成可写的
        printf("s=\"%s\" -> %d (expect %d)\n", cases[i], out, expected[i]);//这里的反斜杠是转义字符
    }
    return 0;
}

File: Algorithm-Templates/LeetCode-Daily/0003-LongestSubstring/LongestSubstring.py
class Solution:
    # LeetCode 标准提交模板，所有题统一包在 Solution 类里
    def lengthOfLongestSubstring(self, s: str) -> int:
        # last: 哈希表，记录每个字符最近一次出现时的下标
        last: dict[str, int] = {}
        left = 0  # 滑动窗口的左边界（包含）
        ans = 0   # 记录当前找到的最长无重复子串的长度

        # right 是窗口右边界下标，ch 是 s[right] 位置的字符
        for right, ch in enumerate(s):
            # 如果 ch 出现过，并且上次出现的位置仍在窗口内
            if ch in last and last[ch] >= left:
                # 把左边界跳到上次出现位置的下一个，把重复字符排除
                left = last[ch] + 1

            # 更新 ch 的最新位置为当前 right
            last[ch] = right

            # 计算当前窗口 [left, right] 的长度
            cand = right - left + 1
            # 如果比已知最长还要长，就更新 ans
            if cand > ans:
                ans = cand

        return ans  # 返回最长子串的长度


# 包装一层普通函数，方便本地测试时不用每次都写 Solution()
def length_of_longest_substring(s: str) -> int:
    """与 Solution 相同逻辑，便于本地单测。"""
    return Solution().lengthOfLongestSubstring(s)


# 仅在直接运行此文件时执行测试，被 import 时不运行
if __name__ == "__main__":
    # (输入字符串, 预期输出)  覆盖了各种情况
    samples = [
        ("abcabcbb", 3),  # 常规重复
        ("bbbbb", 1),     # 全相同字符
        ("pwwkew", 3),    # 重复在中间
        ("", 0),          # 空字符串
    ]
    for s, want in samples:
        got = length_of_longest_substring(s)  # 调用函数
        print(f"s={s!r} -> {got} (expect {want})")
